File: cpp-std/demo17b-reentrant-lock.cpp
/*
REENTRANT LOCKS (RECURSIVE MUTEXES)
Version B: Solving the problem from version A
*/


#include <iostream>
#include <thread>
#include <mutex>
using namespace std;



std::recursive_mutex mut;



void doTask() {
    mut.lock();
    cout << "First time acquiring the resource" << endl;

    mut.lock();
    cout << "Second time acquiring the resource" << endl;

    mut.unlock();
    mut.unlock();
}



int main() {
    auto th = std::thread(doTask);

    th.join();

    return 0;
}

File: csharp/demo/demo17b-reentrant-lock.cs
/*
 * REENTRANT LOCKS (RECURSIVE MUTEXES)
 * Version A: A multithreaded app example
 */
using System;
using System.Collections.Generic;
using System.Threading;



class Demo17B : IRunnable
{
    public void run()
    {
        const int NUM_THREADS = 3;
        var lstWk = new List<Worker>();

        for (int i = 0; i < NUM_THREADS; ++i)
        {
            lstWk.Add(new Worker((char)(i + 'A')));
        }

        lstWk.ForEach(wk => wk.start());
    }



    class Worker
    {
        private static object lk = new object();

        private Thread th;
        private char name;

        public Worker(char name)
        {
            this.name = name;
            this.th = new Thread(doTask);
        }

        private void doTask()
        {
            Thread.Sleep(1000);

            lock (lk)
            {
                Console.WriteLine($"First time {name} acquiring the resource");

                lock (lk)
                {
                    Console.WriteLine($"Second time {name} acquiring the resource");
                }
            }
        }

        public void start()
        {
            th.Start();
        }
    }
}
